_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,1079 @@
+
+```
+(gdb) svd/x SCB BFAR
+Fields in SCB BFAR:
+	BFAR:  0x0BADCAFE  Bus fault address
+
+(gdb) p/a *frame
+$16 = {
+  r0 = 0x1 <g_pfnVectors+1>,
+  r1 = 0x200003c4 <ucHeap+604>,
+  r2 = 0x10000000,
+  r3 = 0xbadcafe,
+  r12 = 0x200001b8 <ucHeap+80>,
+  lr = 0x195 <prvQueuePingTask+52>,
+  return_address = 0x13a <trigger_crash+22>,
+  xpsr = 0x81000000
+}
+
+(gdb) info line *0x13a
+Line 123 of "./cortex-m-fault-debug/main.c" starts at address 0x138 <trigger_crash+20> and ends at 0x13e <trigger_crash+26>.
+
+(gdb) list *0x13a
+0x13a is in trigger_crash (./cortex-m-fault-debug/main.c:123).
+118	  switch (crash_id) {
+119	    case 0:
+120	      illegal_instruction_execution();
+121	      break;
+122	    case 1:
+===> FAULT HERE
+123	      read_from_bad_address();
+124	      break;
+125	    case 2:
+126	      access_disabled_coprocessor();
+127	      break;
+```
+
+Great, so we have pinpointed the exact code which triggered the issue and can now fix it!
+
+{: #coprocessor-fault-example}
+
+### Coprocessor Fault
+
+#### Code
+
+```c
+void access_disabled_coprocessor(void) {
+  // FreeRTOS will automatically enable the FPU co-processor.
+  // Let's disable it for the purposes of this example
+  __asm volatile(
+      "ldr r0, =0xE000ED88 \n"
+      "mov r1, #0 \n"
+      "str r1, [r0]	\n"
+      "dsb \n"
+      "vmov r0, s0 \n"
+      );
+}
+```
+
+#### Analysis
+
+```
+(gdb) break main
+(gdb) continue
+Breakpoint 4, main () at ./cortex-m-fault-debug/main.c:180
+180	  xQueue = xQueueCreate(mainQUEUE_LENGTH, sizeof(unsigned long));
+(gdb) set g_crash_config=2
+(gdb) c
+Continuing.
+
+Program received signal SIGTRAP, Trace/breakpoint trap.
+0x00000218 in my_fault_handler_c (frame=0x20002d80) at ./cortex-m-fault-debug/startup.c:91
+91	  HALT_IF_DEBUGGING();
+```
+
+We can inspect `CFSR` to get a clue about the crash which took place
+
+```
+(gdb) p/x *(uint32_t*)0xE000ED28
+$13 = 0x8200
+(gdb) svd SCB CFSR_UFSR_BFSR_MMFSR
+Fields in SCB CFSR_UFSR_BFSR_MMFSR:
+[...]
+	NOCP:         1  No coprocessor usage fault.
+[...]
+```
+
+We see it was a coprocessor UsageFault which tells us we either issued an instruction to a non-existent or disabled Cortex-M coprocessor. We know the frame contents are valid so we can inspect that to figure out where the fault originated:
+
+```
+(gdb) p/a *frame
+$27 = {
+  r0 = 0xe000ed88,
+  r1 = 0x0 <g_pfnVectors>,
+  r2 = 0x10000000,
+  r3 = 0x0 <g_pfnVectors>,
+  r12 = 0x200001b8 <ucHeap+80>,
+  lr = 0x199 <prvQueuePingTask+52>,
+  return_address = 0x114 <access_disabled_coprocessor+12>,
+  xpsr = 0x81000000
+}
+
+(gdb) disassemble 0x114
+Dump of assembler code for function access_disabled_coprocessor:
+   0x00000108 <+0>:	ldr	r0, [pc, #16]	; (0x11c)
+   0x0000010a <+2>:	mov.w	r1, #0
+   0x0000010e <+6>:	str	r1, [r0, #0]
+   0x00000110 <+8>:	dsb	sy
+===> FAULT HERE
+   0x00000114 <+12>:	vmov	r0, s0
+   0x00000118 <+16>:	bx	lr
+```
+
```

Comment:
Folks may not be familiar with ARM assembly. Might be useful to say "this instruction <instruction> is a floating point instruction".